### PR TITLE
Stripe payment processing

### LINKED ISSUE #7644: Flash error "The payment could not be processed. The payment could not be completed"  when customers provide payment authorization
## Description

Despite payment success,  the customer sees a red flash error indicating that `"The payment could not be processed. The payment could not be completed"`, after using the received link per email to payment provide authorization.

Refreshing the page displays:
`"The payment could not be processed. Invalid payment state: succeeded"`

Note that navigating to other pages displays the error as well - this is the "memory effect" bug #7600.

## Expected Behavior

If the payment is successful the flash error should not be displayed.

## Actual Behaviour

A successfully authorized payment (verified in Stripe) triggers a misleading flash error to the user.

## Steps to Reproduce

As an admin:
1. Create a BO order and add a Stripe SCA payment using the x-3184 card

As a customer:
2. Log in as the respective customer
3. Take the link received per email and paste it into the customers browser session
4. Click "Complete Authentication"
5. See the error `"The payment could not be processed. The payment could not be completed."`
6. Refresh the page.
7. See the error `"The payment could not be processed. Invalid payment state: succeeded"`

## Animated Gif/Screenshot

Peek 2021-05-14 09-53

After refreshing the page, as a customer, the error changes to:

image

## Workaround

Proposing:
bug-s2: a non-critical feature is broken, no workaround

## Your Environment

* Version used: v3.7.1
* Browser name and version: Firefox 88
* Operating System and version (desktop or mobile): Desktop Ubuntu 20.04

## Possible Fix

Maybe related to #7614.

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Closes #7644 (and others)

- Payment Intent validation was returning errors when a payment was successful.
- Payments were not being processed correctly if the order was complete, including with subs
- Various other things. See discussions below.

#### What should we test?

We should test issue #7644 and also check for any regressions on #7509 and check Stripe payment processing in general.

#### Release notes

Fixed a bug where error messages were shown on Stripe payments that were actually successful.

Changelog Category: User facing changes

### CODE REVIEW

File: app/controllers/spree/orders_controller.rb [hunk 1]
```diff
@@ -27,13 +27,7 @@ class OrdersController < ::BaseController
...
       @order = Spree::Order.find_by!(number: params[:id])
 
-      if params.key?("payment_intent")
-        result = ProcessPaymentIntent.new(params["payment_intent"], @order).call!
-        unless result.ok?
-          flash[:error] = "#{I18n.t("payment_could_not_process")}. #{result.error}"
-        end
-        @order.reload
-      end
+      handle_stripe_response
```
Comment (andrewpbrett) [inline app/controllers/spree/orders_controller.rb:30]: We could even make this its own route, and a separate controller action, but this is 90% of the way there as far as clarity.

File: app/models/spree/gateway/stripe_sca.rb [hunk 2]
```diff
@@ -129,7 +129,19 @@ def fetch_payment_intent(creditcard, gateway_options)
...
+          call(payment.response_code, stripe_account_id)
+
+        raise_if_not_in_capture_state(payment_intent_response)
+
+        payment.response_code
+      end
+
+      def raise_if_not_in_capture_state(payment_intent_response)
+        state = payment_intent_response.status
+        return unless state != 'requires_capture'
```
Comment (andrewpbrett) [inline app/models/spree/gateway/stripe_sca.rb:142]:
```suggestion
        return if state == 'requires_capture'
```

**Review by andrewpbrett — APPROVED** [review]
There are some minor things that we could debate about, but given that it's already in production we gotta 🚀 and I think this will fix the issue. I'm going to move it to test ready @filipefurtad0.

File: app/models/spree/gateway/stripe_sca.rb (hunk 2 above)
Comment (Matt-Yorkley) [reply app/models/spree/gateway/stripe_sca.rb:142]: I just cut-and-pasted this from one place to another, but yeah that's much better :+1:

**Comment by Matt-Yorkley** [comment]
@andrewpbrett one thing I was concerned about when looking through this code is the `#can_complete?` methods called here: 

https://github.com/openfoodfoundation/openfoodnetwork/blob/cea3ee4ef9a852936c61585ecd0875ab70f52340/app/services/process_payment_intent.rb#L38

Do you know where it's actually defined? It's not anywhere in our codebase and all the specs that would touch it all stub the response. Are there some dynamic method names here with `ActiveMerchant` or something?

I didn't go into it as it was out of scope for fixing this bug.

**Comment by andrewpbrett** [comment]
That's coming from the state machine: 

Screen Shot 2021-05-14 at 10 14 51 AM

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

Thanks for this PR and adding some specs to this :heart: 

I staged the PR in staging UK and staging-AU, to make sure we are not observing any "server-side-effect". Below are the results for staging-AU, but these overlap with staging-UK.

In staging-AU we had a build staged, in which we do not observe the bug on this issue:

Peek 2021-05-14 20-46

After staging the PR, and refreshing the page, this error message appears (notice, that **this order is paid**):

Peek 2021-05-14 21-08

Also, **after creating a new order** we can see that **backoffice payments are now failing**:

Peek 2021-05-14 21-29

So, it the error message is displayed but expected, but expected in this case. We can clearly see this on the Stripe-Dashboard below:

second line - before this PR
first line - after this PR

image

I did not run thorough Stripe SCA regression tests, as this PR might need a second look I think: it worsens a bit the scenario described on the issue, as backoffice payments now fail. Moving back to in dev.

**Comment by Matt-Yorkley** [comment]
> refreshing the page, this error message appears

I've added a commit to change the way those error messages are persisted between page loads :ok_hand: 

Note: if you're testing that specifically, you might have to reload the page more than once after deploying a new branch though, as an error message could still be stored in the session from the previous deployment.

> it worsens a bit the scenario described on the issue, as backoffice payments now fail

Just to clarify; that's compared to a build from 2 weeks ago as opposed to `master`, right?

**Comment by Matt-Yorkley** [comment]
> So, the error message is displayed, but expected in this case. We can clearly see this on the Stripe-Dashboard below

As far as I can tell, the payment failed to be captured on the _Stripe side_ in that example, and the error shown on our side was correct (because the payment actually wasn't completed by Stripe). If it had succeeded on the Stripe side, I think we would have seen the correct response on our side as well.

So... I'm not sure what the next move is here in terms of debugging this, because the code changes in this PR and not actually changing that behavior on the Stripe side, just the way we handle the response from Stripe after it's finished.

**Comment by filipefurtad0** [comment]
> I've added a commit to change the way those error messages are persisted between page loads ok_hand

Nice! But just to clarify: this test case was aiming to show that after staging the PR (at the time of testing) visiting an order confirmation from a successful order and payment displayed an error, which does not seem right,as the order was previously paid. The test case consisted in:

Before staging the PR (i.e., in the build from two weeks ago)
- following the steps to reproduce in #7644: the payment is successful and no error is observed
- Stage this PR
- Visit that order confirmation or simply refresh the page -> Error appears (in a paid order)

This is a bit different the scenario leading to the persistent flash-errors between page reloads (#7600). I'll re-test this as well, after your newest commit :+1: 

> Just to clarify; that's compared to a build from 2 weeks ago as opposed to master, right?

Yes, that's it :ok_hand:

**Comment by filipefurtad0** [comment]
> So... I'm not sure what the next move is here in terms of debugging this, because the code changes in this PR and not actually changing anything that would cause the payment to either succeed or fail on the Stripe side, just the way we handle the response from Stripe after it's finished. Know what I mean?

Humm, ok thanks for adding this info. I'm a bit puzzled then, because the same order, payment method, etc. was successful before staging (in staging-UK - despite the bug we're aiming to solve; and in staging-AU, with this build 8672969). In both servers, the payment fails after staging this PR... This was observed in staging-AU but was replicated as well on staging-UK as well (yesterday):

Peek 2021-05-14 20-33

Notice that the order is in payment state "Not paid" after authentication; and that the item is still in the cart. Had the order been successful paid it would transition to state "complete" and the cart would have been reset back to zero.

**Comment by Matt-Yorkley** [comment]
So my next question is: what does this scenario look like if we do it with `master`?

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

Repeating this scenario in `master` reproduces bug #7644: it leads to a successful payment, but displays the error messages described in the issue.

The release branch is currently [staged](https://semaphoreci.com/openfoodfoundation/openfoodnetwork-2/branches/master/builds/2578) in staging-FR:

Peek 2021-05-16 10-14

**Comment by Matt-Yorkley** [comment]
Aha! Okay, I think I've got it now. Could you try with this new version?

**Comment by Matt-Yorkley** [comment]
In that new example above, what does the Stripe Dashboard say about whether or not that payment has been completed on their side?

**Comment by Matt-Yorkley** [comment]
Okay! I think it's all fixed up now...

**Comment by filipefurtad0** [comment]
Alright! Just reproduced the scenario, after staging these changes:

Peek 2021-05-16 16-12

The main points:

- no error message, after the successful payment
- the order is marked as "Paid" (customer)
- the order is in state "Complete" (backoffice)
- the payment is appears as captured/succeeded on the Stripe Dashboard

This pretty much demonstrates that the PR closes issue #7644 - awesome :tada: 

As you pointed before, it's probably a good idea to run a sanity check on Stripe, and specifically on #7509 (Stripe + subs). I'll drop a line again as soon as I've gone through these test cases :+1:

**Comment by filipefurtad0** [comment]
Ran the following tests:

1) Backoffice orders

i) with **valid** card x-3184; refusing authentication
Outcome: error message (customer); order not complete, not paid (verified on customer, admin and stripe) - OK

ii) with **invalid** card x-3178; completing authentication
Outcome: "Insufficient funds" error message (customer); order not complete, not paid (verified on customer, admin and stripe) - OK

2) Shopfront orders

i) with **valid** card x-3184; completing authentication
Outcome: payment is successful, order changes state from "payment" to "complete - OK

ii) with **valid** card x-3184; refusing authentication
Outcome: error "The provided PaymentMethod has failed authentication. You can provide payment_method_data or a new PaymentMethod to attempt to fulfill this PaymentIntent again." is displayed; order remains in "payment" state (not "complete")

iii) with **invalid** card x-3178; refusing authentication
Outcome: "Your card has insufficient funds." error message (customer); order remains in "payment" state (not "complete")

3) Subscription orders

Placing a subscription having granted the hub with authorization for:
i) a **valid** card x-3184; completing authentication - **failing** -> this is expected to succeed
- The flash error "The payment could not be completed" is displayed
- the order remains "not paid" (customer)
- it is in state "complete" despite of failed payment (this relates to issue #7510, which sprang the discussion on an additional payment state for these cases, but is out of scope here - something to chat about)
- the payment state is "balance due"

See:

Peek 2021-05-16 18-12


ii) a **valid** card x-3184; failing authentication - OK
Same as in 2.ii) - flash error message: "The payment could not be processed. The provided PaymentMethod has failed authentication. You can provide payment_method_data or a new PaymentMethod to attempt to fulfill this PaymentIntent again. "; payment in "balance due" state, as expected.

iii) an **invalid** card x-3178; completing authentication - OK
same as 2.iii) - flash error message "The payment could not be processed. Your card has insufficient funds. "; payment in "balance due" state as expected.

iv) a **valid** non-3D-auth card x-4242; completing authentication - OK
payment goes through, order is in state "complete" and payment state is "paid".


Ok, so the **summary from these test cases is**:

1) Backoffice orders - all good
2) Shopfront orders - all good
3) Subscription orders - case i) is failing

- the payment with a valid 3D-auth card (authentication completed) of a subscription order - test-case 3.i) - which is expected to succeed, but is failing. I've tried setting up the "fresh card", but still can't get the order to be paid...

Here's the log from Stripe, in case this is helpful:

image

**Comment by Matt-Yorkley** [comment]
@filipefurtad0 Ah I was going to ask about a similar case, as I thought it would probably have problems, which is: when the order is already is "compete" state and we try to process a payment (like your failed example above for subs). Does that work currently in `master`? When we process a payment on a subscription like that, is the order always in "compete" state before we authorize the payment?

**Comment by Matt-Yorkley** [comment]
I think we can encounter a similar case if an order was previously compete and paid for, then some more items are added to the order. In that case the payment status will be "payment due". If we then process a second payment on the competed order, I think there will be problems. There's a bit of a hole there in both the payment processing logic and the specs...

**Comment by filipefurtad0** [comment]
> When we process a payment on a subscription like that, is the order always in "compete" state before we authorize the payment?

Yes, that's the case for subscription orders, which use Stripe-SCA as a payment method: these are in state "complete" before an actual payment authorization/authentication attempt - #7510.

**Comment by filipefurtad0** [comment]
> I think we can encounter a similar case if an order was previously competed and paid for, then some more items are added to the order. In that case the payment status will be "payment due". If we then process a second payment on the competed order, I think there will be problems. 

This sounds like a rather relevant scenario and test case :+1: I'm not entirely sure to have tried this in master...

**Comment by filipefurtad0** [comment]
> when the order is already is "compete" state and we try to process a payment (like your failed example above for subs). Does that case work currently in master? 

This should be working in master, yes :+1: (I'll make a quick check)

I'm not so sure about that case with more that one payment (due to stock changes)... Something to check.

**Comment by Matt-Yorkley** [comment]
> These payments from subscription orders are never "captured" on Stripe, but I think this is expected, right?

That means the payment hasn't actually been completed on the Stripe side, but it's marked as competed on our side, right? The difference with this PR is that it's correctly showing that the payment hasn't been captured in that case...

**Comment by Matt-Yorkley** [comment]
Unless I've missed something? Those "completed" payments above can't be manually captured, right?

**Comment by filipefurtad0** [comment]
> The difference with this PR is that it's correctly showing that the payment hasn't been captured in that case...

I see :+1: 

> Those "completed" payments above can't be manually captured from the UI, right?

We cannot capture them from the UI. And I think manually capturing them on Stripe would probably not update the orders to the correct state... (with this PR, I'd have to test that). If this is the case, then these orders would remain appearing as "not paid"/"balance due" which could lead to additional payment attempts.

So, I guess in that case the missing bit is making these completed SCA payments captured automatically, without the need for further action on Stripe nor the UI. Would this be the correct/expected behavior?

**Comment by Matt-Yorkley** [comment]
Alright, I think those payments will now be captured correctly if the order is in competed state (either with subs or with secondary payments on changed orders), and I think this issue should be fixed as well: 

> by "mixing" payment methods: first adding an SCA payment (not authorizing it); then adding and capturing a cash payment; after that authorizing the SCA payment triggers this error

**Comment by filipefurtad0** [comment]
Awesome @Matt-Yorkley - The latest commits clearly address the **previously failing test-case (3.i)**: a subs order paid with a StripeSCA card; this is now succeeding :tada: 

Peek 2021-05-17 09-47

Points to notice:
- the payment is automatically captured on Stripe, with no need for manual action
- as such, the payment state on the respective order is marked as "paid" both on admin and customer's side

I felt it was a good idea to do a quick check on **3.ii) - not providing payment/authentication**, regarding the previous bug on #7509:
- the order is marked as "not paid" (customer)
- the payment is in state "balance due" (admin)
- the payment remains marked as "incomplete" (Stripe Dashboard)

We're good here!

As to the **"mixing payments" test case**, there perhaps several scenarios here worth exploring. To keep this under scope and unblock the release asap I addressed this one only:

- creating a BO order
- adding a stripe-SCA payment (but not providing authorization)
- adding a cash payment fully covering the order total -> "Error saving payment" message is displayed in the UI
- refreshing the page shows that the cash payment was accepted though :+1: 
- capturing the cash payment works -> the order does not immediately proceed to "complete" state; the page needs refreshing
- provide authorization for the payment -> no error is triggered :+1: 
- the Stripe payment is displayed on the Stripe dashboard as "uncaptured"; it is still possible to capture this payment manually, but I think this is fine?

We can maybe open an issue do address regarding the UI error, but I can confirm the previous bugsnag error was not triggered :+1: 

Let's merge this! This is a major improvement - thank you! :tada: 
Moving to ready to go.

**Comment by Matt-Yorkley** [comment]
Thanks @filipefurtad0 !! :heart:

File: spec/models/spree/gateway/stripe_sca_spec.rb [hunk 3]
```diff
@@ -50,6 +50,35 @@
       expect(response.success?).to eq false
       expect(response.message).to eq "Invalid payment state: succeeded"
     end
+
+    context "when payment intent state is not in 'requires_capture' state" do
+      before do
+        payment
+      end
+
+      it "succeeds if payment intent state is requires_capture" do
```
Comment (sauloperez) [inline spec/models/spree/gateway/stripe_sca_spec.rb:59]: Personally, when I spot this pattern `it "<action> if/when <condition>` to a

```rb
context `when <condition>`
  it `<action>`
  end
end
```

IMO this improves readability as I'm able to skim through `contexts` and map those into conditionals in my head. I know the code under test is expected not to behave always the same way. Also, I believe this approach also leverages RSpec's DSL.

File: app/controllers/spree/orders_controller.rb [hunk 4]
```diff
@@ -164,7 +164,7 @@ def handle_stripe_response
       result = ProcessPaymentIntent.new(params["payment_intent"], @order).call!
 
       unless result.ok?
-        flash[:error] = "#{I18n.t("payment_could_not_process")}. #{result.error}"
+        flash.now[:error] = "#{I18n.t("payment_could_not_process")}. #{result.error}"
```
Comment (sauloperez) [inline app/controllers/spree/orders_controller.rb:167]: finally! It was for sure something stupid :sweat:

File: app/services/process_payment_intent.rb [hunk 5]
```diff
@@ -35,8 +35,7 @@ def call!
 
     OrderWorkflow.new(order).next
 
-    if last_payment.can_complete?
-      last_payment.complete!
```
Comment (sauloperez) [inline app/services/process_payment_intent.rb:39]: so the OrderWorkflow completes it already, right?

File: app/services/process_payment_intent.rb [hunk 6]
```diff
@@ -50,25 +49,36 @@ def call!
...
-  def valid?
-    order.present? && matches_last_payment?
+  def process_payment
+    if order.state == "payment"
+      # Moves the order to completed, which calls #process_payments!
+      OrderWorkflow.new(order).next
+    else
+      order.process_payments!
+    end
+  end
```
Comment (sauloperez) [inline app/services/process_payment_intent.rb:61]: This is IMO flagging a flaw. So far it was OrderWorkflow the one responsible for this, and this change leaks this concern. As a result, there are now two places we will need to check next time we're investigating or developing something related to payment processing. I would suggest we create a tech debt issue and carefully investigate this design problem without any rush.

I think the underlying issue here is having all payment processing logic hanging from the order model. I always knew this wasn't a good idea and I think now I have proof.

File: app/services/process_payment_intent.rb (hunk 5 above)
Comment (Matt-Yorkley) [reply app/services/process_payment_intent.rb:39]: Yes and no! If the order is in payment state, `OrderWorkflow.new(order).next` will move it to completion, which processes the payments, and sets the payment to complete (if capturing was successful).

In the case that the order is already completed, `OrderWorkflow.new(order).next` will not actually process the payments correctly!

File: app/services/process_payment_intent.rb (hunk 6 above)
Comment (Matt-Yorkley) [reply app/services/process_payment_intent.rb:61]: Agreed it needs a lot of re-thinking! The issue here is `#process_payments!` only gets called on the final state transition to complete, and there are lots of cases (including subscriptions) where the order is already compete when we want to process the payment via Stripe.

Comment (sauloperez) [reply app/services/process_payment_intent.rb:61]: I see your comment on line 56 and it makes sense to me now. It says `OrderWorkflow` does two things: moving to completion && calling process payments => this violates SRP. If these two weren't glued together we could choose and process payments always, regardless of the order state.

Comment (Matt-Yorkley) [reply app/services/process_payment_intent.rb:61]: A lot of the issues here come from the fact that the mechanics of the checkout order placement flow and the admin order placement flow are totally different, but we try to re-use the same logic as if there are no differences. There's lots of things that need to be decoupled and updated to make it work nicely with both!

File: app/services/process_payment_intent.rb (hunk 5 above)
Comment (sauloperez) [reply app/services/process_payment_intent.rb:39]: I see. That's clearly a leaky abstraction.

**Comment by sauloperez** [comment]
None of my comments block this. They just food for thought. I think we're seeing more and more the limits of the existing design.

**Review by sauloperez — APPROVED** [review]

**Pull request merged**